Assets/Scripts/Map/MapPatrollingGen/WatchmanRouteSolver.cs

```diff
@@ -0,0 +1,417 @@
+                closestVertexInNew.AddNeighbor(closestVertexInCurrent);
+
+                // Merge the closest island into the current island
+                currentIsland.AddRange(closestIsland);
+                remainingIslands.Remove(closestIsland);
+
+                // Recursive call to merge the next closest island
+                return MergeIslandsRecursively(currentIsland, remainingIslands, distanceDict);
+            }
+
+            return currentIsland; // Return the current island if no more connections can be made
+        }
+
+        private static void ConnectIslands(
+            Vertex[] vertices,
+            Dictionary<(Vector2Int, Vector2Int), int> distanceDict)
+        {
+            var visited = new HashSet<Vertex>();
+            var components = new List<List<Vertex>>();
+
+            // Identify all connected components using DFS
+            foreach (var vertex in vertices)
+            {
+                if (!visited.Contains(vertex))
+                {
+                    var component = new List<Vertex>();
+                    DepthFirstSearch(vertex, visited, component);
+                    components.Add(component);
+                }
+            }
+
+            if (components.Count > 1)
+            {
+                // Start with the first component as the initial island
+                var initialIsland = components[0];
+                components.RemoveAt(0);
+
+                // Recursively merge all components into one island
+                MergeIslandsRecursively(initialIsland, components, distanceDict);
+            }
+        }
+
+        // Calculate the shortest path between all pairs of verticies
+        private static Dictionary<(Vector2Int, Vector2Int), int> CalculateDistanceMatrix(bool[,] map, List<Vector2Int> verticies)
+        {
+            Dictionary<(Vector2Int, Vector2Int), int> shortestGridPath = new();
+
+            foreach (var vertex in verticies)
+            {
+                BreathFirstSearch(vertex, shortestGridPath, verticies, map);
+            }
+
+            return shortestGridPath;
+        }
+
+        // Function to check if a position is within bounds and walkable
+        private static bool IsWalkable(Vector2Int pos, bool[,] map)
+        {
+            return pos.x >= 0 && pos.x < map.GetLength(0) &&
+                   pos.y >= 0 && pos.y < map.GetLength(1) &&
+                   !map[pos.x, pos.y]; // true if the position is not a wall
+        }
+
+        // BFS to find the shortest path from the start position to all other verticies
+        private static void BreathFirstSearch(Vector2Int startPosition, Dictionary<(Vector2Int, Vector2Int), int> shortestGridPath, List<Vector2Int> guardPositions, bool[,] map)
+        {
+            var directions = new Vector2Int[]
+            {
+            new(0, 1),
+            new(0, -1),
+            new(1, 0),
+            new(-1, 0)
+            };
```

Comment:
Move this out to a private static readonly Vector2Int[] so we don't have to allocate an array every time.